src/Futhark/CodeGen/Backends/GenericWASM.hs

```diff
@@ -0,0 +1,400 @@
+extToString (Imp.TransparentValue _ (Imp.ScalarValue (FloatType Float32) _ _)) = "f32"
+extToString (Imp.TransparentValue _ (Imp.ScalarValue (FloatType Float64) _ _)) = "f64"
+extToString (Imp.TransparentValue _ (Imp.ScalarValue (IntType Int8) Imp.TypeDirect _)) = "i8"
+extToString (Imp.TransparentValue _ (Imp.ScalarValue (IntType Int16) Imp.TypeDirect _)) = "i16"
+extToString (Imp.TransparentValue _ (Imp.ScalarValue (IntType Int32) Imp.TypeDirect _)) = "i32"
+extToString (Imp.TransparentValue _ (Imp.ScalarValue (IntType Int64) Imp.TypeDirect _)) = "i64"
+extToString (Imp.TransparentValue _ (Imp.ScalarValue (IntType Int8) Imp.TypeUnsigned _)) = "u8"
+extToString (Imp.TransparentValue _ (Imp.ScalarValue (IntType Int16) Imp.TypeUnsigned _)) = "u16"
+extToString (Imp.TransparentValue _ (Imp.ScalarValue (IntType Int32) Imp.TypeUnsigned _)) = "u32"
+extToString (Imp.TransparentValue _ (Imp.ScalarValue (IntType Int64) Imp.TypeUnsigned _)) = "u64"
+extToString (Imp.TransparentValue _ (Imp.ScalarValue Bool _ _)) = "bool"
+extToString (Imp.OpaqueValue _ oname vds) = opaqueName oname vds
+extToString ev = error $ "extToString: missing case: " ++ show ev
+
+type EntryPointTyp = String
+
+data JSEntryPoint = JSEntryPoint
+  { name :: String,
+    parameters :: [EntryPointTyp],
+    ret :: [EntryPointTyp]
+  }
+
+emccExportNames :: [JSEntryPoint] -> [String]
+emccExportNames jses =
+  map (\jse -> "'_futhark_entry_" ++ name jse ++ "'") jses
+    ++ map (\arg -> "'" ++ gfn "new" arg ++ "'") arrays
+    ++ map (\arg -> "'" ++ gfn "free" arg ++ "'") arrays
+    ++ map (\arg -> "'" ++ gfn "shape" arg ++ "'") arrays
+    ++ map (\arg -> "'" ++ gfn "values_raw" arg ++ "'") arrays
+    ++ map (\arg -> "'" ++ gfn "values" arg ++ "'") arrays
+    ++ map (\arg -> "'" ++ "_futhark_free_" ++ arg ++ "'") opaques
+    ++ [ "_futhark_context_config_new",
+         "_futhark_context_config_free",
+         "_futhark_context_new",
+         "_futhark_context_free",
+         "_futhark_context_get_error"
+       ]
+  where
+    arrays = filter isArray typs
+    opaques = filter isOpaque typs
+    typs = nub $ concatMap (\jse -> parameters jse ++ ret jse) jses
+    gfn typ str = "_futhark_" ++ typ ++ "_" ++ baseType str ++ "_" ++ show (dim str) ++ "d"
+
+javascriptWrapper :: [JSEntryPoint] -> String
+javascriptWrapper entryPoints =
+  unlines
+    [ jsServer,
+      jsValues,
+      classFutharkOpaque,
+      classFutharkArray,
+      classFutharkContext entryPoints
+    ]
+
+jsServer :: String
+jsServer = $(embedStringFile "rts/javascript/server.js")
+
+jsValues :: String
+jsValues = $(embedStringFile "rts/javascript/values.js")
+
+classFutharkOpaque :: String
+classFutharkOpaque =
+  T.unpack
+    [text|
+  class FutharkOpaque {
+    constructor(ctx, ptr, ffree) { this.ctx = ctx; this.ptr = ptr; this.ffree = ffree; }
+    free() { this.ffree(this.ctx, this.ptr); }
+  }
+  Module['FutharkOpaque'] = FutharkOpaque;
+  |]
+
+classFutharkArray :: String
+classFutharkArray =
```

Comment:
Could this also be moved into a file in `rts/javascript`?

Reply:
The implementation of the FutharkArray and FutharkOpaque class can be moved to `rts/javascript`. Should I just make two files `array.js` and `opaque.js`. Or should they be one file? and are the suggested names of the files ok?

Reply:
I'd put them in the same file.  It doesn't much matter much what it's called.  After five seconds of thought, I'd call it `wrapperclasses.js`.